docs/proposals/mapper-design.md

```diff
@@ -0,0 +1,117 @@
+---
```

Comment:
overall design looks good. needs more refinement. we can keep on improving the details.
@qizha , please review.